### ISSUE TITLE
Require initialValue for .reduce()

### ISSUE THREAD

**Issue by fkereki** [eslint/eslint#12868]
I'm reposting this from a proposal I found at the TSLINT project, at https://github.com/palantir/tslint/issues/4173. I have found problems in a code base myself, which would have been avoided if this rule had been in place, so I second the proposal.

**Please describe what the rule should do:**
Requires `initialValue` to be specified when calling `Array.prototype.reduce()`

When using `Array.prototype.reduce()` the `initialValue` parameter is optional. As per MDN docs: "If the array is empty and no `initialValue` is provided, TypeError will be thrown." Therefore the suggestion is to provide a rule where the initialValue is required to be specified to avoid TypeErrors being raised.

**What category of rule is this? (place an "X" next to just one item)**

[X] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Enforces code style (layout)
[ ] Other (please specify:)

It is a general problem that is likely to occur as the code grows.

**Provide 2-3 code examples that this rule will warn about:**

As per MDN docs:
"It is usually safer to provide an initial value because there are three possible outputs without initialValue, as shown in the following example."

```js
var maxCallback = ( acc, cur ) => Math.max( acc.x, cur.x );
var maxCallback2 = ( max, cur ) => Math.max( max, cur );

// reduce() without initialValue
[ { x: 22 }, { x: 42 } ].reduce( maxCallback ); // 42
[ { x: 22 }            ].reduce( maxCallback ); // { x: 22 }
[                      ].reduce( maxCallback ); // TypeError

// map/reduce; better solution, also works for empty or larger arrays
[ { x: 22 }, { x: 42 } ].map( el => el.x )
                        .reduce( maxCallback2, -Infinity );
```
Another example: when reduce is being called on an array passed in as a parameter, or injected (e.g., using Redux), providing the `initialValue` parameter to reduce avoids having to check if the array is empty, however, that is easy to forget while coding.

```js
function sumArray(values: number[]) {
    return values.reduce((a, b) => a + b);
}

sumArray([1,2]); // 3
sumArray([]); // TypeError

function sumArrayWithInitial(values: number[]) {
    return values.reduce((a, b) => a + b, 0);
}

sumArrayWithInitial([]); // 0
```

**Why should this rule be included in ESLint (instead of a plugin)?**

This is a generic JavaScript rule, not linked to frameworks or libraries.

**Are you willing to submit a pull request to implement this rule?**

Don't have enough knowledge about writing rules, so "no" at least at this time.

**Comment by mdjermanovic** [comment]
Hi @fkereki, thanks for the new rule proposal!

I think this would be a useful rule, but its scope might be too narrow for a core rule since it's targeting only one specific method.

If the rule doesn't reach [consensus](https://eslint.org/docs/maintainer-guide/issues#accepting-issues) to be included in the core, an alternative may be to configure the generic [no-restricted-syntax](https://eslint.org/docs/rules/no-restricted-syntax) rule to catch this:

```js
/* eslint "no-restricted-syntax": ["error",
  {
    "selector": "CallExpression[arguments.length=1] > MemberExpression.callee > Identifier.property[name='reduce']",
    "message": "Provide initialValue to .reduce()."
  }
]
*/        

arr.reduce(f); // error

values.reduce((a, b) => a + b); // error

values.reduce((a, b) => a + b, 0); // ok
```

[Online Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IFwibm8tcmVzdHJpY3RlZC1zeW50YXhcIjogW1wiZXJyb3JcIixcbiAge1xuICAgIFwic2VsZWN0b3JcIjogXCJDYWxsRXhwcmVzc2lvblthcmd1bWVudHMubGVuZ3RoPTFdID4gTWVtYmVyRXhwcmVzc2lvbi5jYWxsZWUgPiBJZGVudGlmaWVyLnByb3BlcnR5W25hbWU9J3JlZHVjZSddXCIsXG4gICAgXCJtZXNzYWdlXCI6IFwiUHJvdmlkZSBpbml0aWFsVmFsdWUgdG8gLnJlZHVjZSgpLlwiXG4gIH1cbl1cbiovICAgICAgICBcblxuYXJyLnJlZHVjZShmKTsgLy8gZXJyb3JcblxudmFsdWVzLnJlZHVjZSgoYSwgYikgPT4gYSArIGIpOyAvLyBlcnJvclxuXG52YWx1ZXMucmVkdWNlKChhLCBiKSA9PiBhICsgYiwgMCk7IC8vIG9rIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo2LCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7ImNvbnNpc3RlbnQtcmV0dXJuIjoyfSwiZW52Ijp7fX19)

**Comment by kaicataldo** [comment]
I agree that this would be a useful rule, but I echo @mdjermanovic's thoughts above. We have a very small volunteer-based team and have a lot of rules to maintain. We have to set a very high bar for new rules in core, and as this can be covered by `no-restricted-syntax`, I'm not sure it meets that bar. Let's see what others on the team have to say.

**Comment by g-plane** [comment]
Will there be false positive when a user-created object which has a method also named "reduce"? ESLint can't get type information.

**Comment by kaicataldo** [comment]
> Will there be false positive when a user-created object which has a method also named "reduce"? ESLint can't get type information.

It would, but we also have other rules that have similar false positives like [`array-callback-return`](https://eslint.org/docs/rules/array-callback-return).

**Comment by g-plane** [comment]
It seems that we can implement this rule in a separated plugin.

**Comment by anikethsaha** [comment]
I think it may be good candidate for [eslint-plugin-fp](https://github.com/jfmengels/eslint-plugin-fp). ?

**Comment by g-plane** [comment]
Maybe.
